libvuln/jsonblob/jsonblob.go

```diff
@@ -117,6 +144,54 @@ func (l *Loader) Entry() *Entry {
 	return l.e
 }
 
+// Iter returns an iterator for read all objects for the current update operation.
+func (l *Loader) Iter() *iter {
```

Comment:
> This shouldn't return an unexported type.

My goal is to prevent code from outside the package to create `iter` objects. I need clarification of why this should not be an unexported type.

Reply:
> Alternatively, maybe just adding a StreamingLoader or IterLoader with a better/different interface instead of intertwining this with Loader would prevent the possibility of misuse.

That's a valid point.

Considering the `Loader` already uses the `Next()/Entry()/Err()` pattern, and considering we don't want to break backward compatibility, it seems less effort to continue using the `Loader` internals vs. a new type. We would want the existing `Loader` to use the `StreamingLoader` anyway (if not, it's one additional code path for this, and I would rather avoid).

So we have the current proposal, where the `Loader` supports two iteration patterns:

1. Calling `Next()` + `Entry()` to retrieve batches per updater.
2. Calling `NextIter()` + `Iter()` to retrieve individual items per updater.

Or your proposal:

> It seems like it should be returning the equivalent of iter.Seq2[Entry, err], documenting that the returned Entry objects contain a single Enrichment or Vulnerability, and calling (*Loader).Iter() is mutually exclusive with calling (*Loader).Entry().

Let's ignore `rangefunc` for now -- which I don't think it worth doing anyway.

That option implies we would reuse `Next()` for both (1.) and (2.) in the previous proposal. It lets the user call `Next()` once and decide if either return the whole batch with `Entry()` or an iterator using `Iter()`. But, that's not possible unless I change the contract of `Entry()` dictating the data is already available when `Next() == true`. I would have to either return `nil` (which I don't think the current contract allows) or partial data (which could result in misuse before the caller checks things failed calling `Loader.Err()`).

A third option is to pass an option to the `Loader`, which would change the behavior of `Next()` and `Entry()` to be of `NextIter() ` and `Iter()` (we would reuse `Entry` to iterate over individual items instead of populating the slices). That's convoluted.

So, the two iterator patterns (the current approach) balance out the need for functionality, minimal changes, and test coverage. I want to maintain backward compatibility, avoid regressions, and add a new access pattern, at the expense of adding two access patterns to the `Loader` API. But I think that's OK because that API will die soon.

Reply:
Quoting myself:

> Let's ignore rangefunc for now -- which I don't think is worth doing anyway.

I lied. I played a bit with the idea today. The implementation turned out simpler, albeit a bit unusual, because I went ahead and used nested iterators. @hdonnay, let me know if you would prefer to pursue that instead https://github.com/quay/claircore/commit/d2531cb8f1fe63af5b2479563e9899f303c61aba

By the way, we would have to keep the previous `Next()/Entry()/Err()` implementation as is for now, so there would be two paths for the same thing.